java-checks/src/main/java/org/sonar/java/checks/helpers/IntersectAutomataChecker.java

```diff
@@ -31,13 +35,20 @@ protected boolean neutralAnswer() {
 
   @Override
   protected boolean checkAuto1AndAuto2Successors(SubAutomaton auto1, SubAutomaton auto2, boolean defaultAnswer, boolean hasConsumedInput) {
-    SimplifiedRegexCharacterClass characterClass1 = SimplifiedRegexCharacterClass.of(auto1.start);
-    SimplifiedRegexCharacterClass characterClass2 = SimplifiedRegexCharacterClass.of(auto2.start);
-    return ((characterClass1 != null) && (characterClass2 != null)) ?
-      (characterClass1.intersects(characterClass2, defaultAnswer) &&
-        auto1.anySuccessorMatch(successor1 -> auto2.anySuccessorMatch(successor2 ->
-          check(successor1, successor2, true)))) :
-      defaultAnswer;
+    boolean start1IntersectsStart2;
+    if (auto1.incomingTransitionType() == BOUNDARY && auto2.incomingTransitionType() == BOUNDARY) {
+      start1IntersectsStart2 = ((BoundaryTree) auto1.start).type() == ((BoundaryTree) auto2.start).type();
+    } else {
+      SimplifiedRegexCharacterClass characterClass1 = SimplifiedRegexCharacterClass.of(auto1.start);
+      SimplifiedRegexCharacterClass characterClass2 = SimplifiedRegexCharacterClass.of(auto2.start);
+      if (characterClass1 == null || characterClass2 == null) {
+        return defaultAnswer;
+      }
+      start1IntersectsStart2 = characterClass1.intersects(characterClass2, defaultAnswer);
+    }
+    return start1IntersectsStart2 &&
```

Comment:
I think the method would be a bit nicer to read if we moved the previous if-else into its own method `charactersIntersect` and replaced `start1IntersectsStart2` with `charactersIntersect(auto1.start, auto2.start)` here.

This would also fix a bug I've just noticed where, if one of the character classes is null and the `defaultAnswer` is `true`, we return true without checking the successors.